src/main/java/org/candlepin/dto/shim/SubscriptionInfoTranslator.java

```diff
@@ -0,0 +1,130 @@
+ * in this software or its documentation.
+ */
+package org.candlepin.dto.shim;
+
+import org.candlepin.dto.ModelTranslator;
+import org.candlepin.dto.ObjectTranslator;
+import org.candlepin.dto.api.v1.CdnDTO;
+import org.candlepin.dto.api.v1.CertificateDTO;
+import org.candlepin.dto.api.v1.NestedOwnerDTO;
+import org.candlepin.dto.api.v1.ProductDTO;
+import org.candlepin.dto.api.v1.SubscriptionDTO;
+import org.candlepin.service.model.ProductInfo;
+import org.candlepin.service.model.SubscriptionInfo;
+import org.candlepin.util.Util;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+
+/**
+ * The SubscriptionTranslator provides translation from SubscriptionInfo objects to the
+ * SubscriptionDTO (hosted test API)
+ */
+public class SubscriptionInfoTranslator implements ObjectTranslator<SubscriptionInfo, SubscriptionDTO> {
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public SubscriptionDTO translate(SubscriptionInfo source) {
+        return this.translate(null, source);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public SubscriptionDTO translate(ModelTranslator translator, SubscriptionInfo source) {
+        return source != null ? this.populate(translator, source, new SubscriptionDTO()) : null;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public SubscriptionDTO populate(SubscriptionInfo source, SubscriptionDTO destination) {
+        return this.populate(null, source, destination);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public SubscriptionDTO populate(
+        ModelTranslator translator, SubscriptionInfo source, SubscriptionDTO dest) {
+        if (source == null) {
+            throw new IllegalArgumentException("source is null");
+        }
+
+        if (dest == null) {
+            throw new IllegalArgumentException("dest is null");
+        }
+
+        dest.id(source.getId())
+            .lastModified(Util.toDateTime(source.getLastModified()))
+            .startDate(Util.toDateTime(source.getStartDate()))
+            .endDate(Util.toDateTime(source.getEndDate()))
+            .accountNumber(source.getAccountNumber())
+            .contractNumber(source.getContractNumber())
+            .orderNumber(source.getOrderNumber())
+            .quantity(source.getQuantity())
+            .upstreamConsumerId(source.getUpstreamConsumerId())
+            .upstreamEntitlementId(source.getUpstreamEntitlementId())
+            .upstreamPoolId(source.getUpstreamPoolId());
+
+        if (translator != null) {
+            dest.cdn(translator.translate(source.getCdn(), CdnDTO.class))
+                .certificate(translator.translate(source.getCertificate(), CertificateDTO.class))
+                .owner(translator.translate(source.getOwner(), NestedOwnerDTO.class));
+            ProductInfo product = source.getProduct();
+            if (product != null) {
```

Comment:
Line break before the if statement

Also need to set product to null on the outbound DTO in the no-translator and no-product cases.

Reply:
Do we need to be explicit for properties other than collections? Objects will just default to null.

Reply:
The populate method is supposed to be idempotent, and is used for both cases where we're translating and populating. That's why we want to set everything to a known state based on the source object.

Reply:
Done.